packages/calling-stateful-client/src/CallContext.ts

```diff
@@ -400,6 +402,59 @@ export class CallContext {
     });
   }
 
+  /* @conditional-compile-remove(reaction) */
+  public setReceivedReactionFromParticipant(
+    callId: string,
+    participantKey: string,
+    reactionMessage: ReactionMessage | null
+  ): void {
+    this.modifyState((draft: CallClientState) => {
+      const call = draft.calls[this._callIdHistory.latestCallId(callId)];
+
+      if (call) {
+        const participant = call.remoteParticipants[participantKey];
+
+        if (participant) {
+          if (reactionMessage == null) {
+            participant.reactionState = undefined;
+          } else {
+            if (
+              call.capabilitiesFeature?.capabilities.reaction &&
+              isReactionBeingPlayedNow(participant.reactionState?.receivedAt)
+            ) {
+              participant.reactionState = undefined;
+              clearTimeout(this._timeOutId[participantKey]);
+            }
+            participant.reactionState = reactionMessage
+              ? { reactionMessage: reactionMessage, receivedAt: new Date() }
+              : undefined;
+            this._timeOutId[participantKey] = setTimeout(() => {
+              clearParticipantReactionState(this, callId, participantKey);
+            }, 2170);
+          }
+        } else if (participantKey === toFlatCommunicationIdentifier(this._state.userId)) {
+          if (reactionMessage == null) {
+            call.localParticipantReactionState.reactionState = undefined;
+          } else {
+            if (
+              call.capabilitiesFeature?.capabilities.reaction &&
+              isReactionBeingPlayedNow(call.localParticipantReactionState.reactionState?.receivedAt)
+            ) {
+              clearParticipantReactionState(this, callId, participantKey);
```

Comment:
nit: is this necessary? We are immediately setting the reactionState to a new value after this `if`.

Reply:
If it isn't we can start to simplify this function, there's a lot of nesting (7+levels) and quite a bit of duplicate code.

Perhaps something like (**this is untested pseudocode**):
```ts
public setReceivedReactionFromParticipant(...)
...modifyState...
{
  const call = draft.calls[this._callIdHistory.latestCallId(callId)];
  if (!call) {
    return;
  }

  // New reaction has been received, clear any existing reaction timeout triggers
  clearTimeout(this._timeOutId[participantKey]);

  // Set the new reaction state on the appropriate participants
  const newReactionState = reactionMessage ? { reactionMessage: reactionMessage, receivedAt: new Date() } : undefined;
  const participant = call.remoteParticipants[participantKey];
  if (paticipantIsRemoteParticipant(participantId) {
    participant.reactionState = newReactionState;
  } else if (participantIsLocalParticipant(participantId)) {
    call.localParticipantReactionState = newReactionState;
  }

  // Set the new reaction state to clear after a timeout
  reactionMessage && this._timeOutId[participantKey] = setTimeout(() => {
    clearParticipantReactionState(this, callId, participantKey);
  }, 2170);
}
```

Reply:
Line 443 allows us to replace the ongoing reaction that was received for rendering. 
This would allow the contoso user to send a corrected reaction if in critical situation a wrong reaction was sent by mistake which might send the wrong sensitivity in the meeting.

Reply:
Thanks for the suggested refactoring. I understand the current case handlings are not the best looking code and can be improved. 
We'll have more changes in this stub with capability check only for teams meeting but not for group calls. Right now we only check if capability exists or not! 
The cases missing in the suggested refactoring is that - 
* when we are doing - clearTimeOut(...) 
`// New reaction has been received, clear any existing reaction timeout triggers
  clearTimeout(this._timeOutId[participantKey]);`
  We need to get update the state as well to clear the current rendering, otherwise videoTile will just render another reaction (the newly received one) on top the old one rather than replacing it. 
  To clear the state, we need to know if the participant is a remote one or a local one. Thus I am clearing them under the check of detecting the participant type.

Reply:
Refactored the stub. Tested manually as well.